### DISCUSSION TITLE
Calling a function based on task result

### DISCUSSION THREAD

**Discussion by georgev93** [stevearc/overseer.nvim#380] [Q&A] [answered]
I'm struggling a bit here due to my lack of understanding of tasks vs. components vs. actions, etc.

I have successfully set up a task for running unit tests, but I want the cursor color to change based on the result of the unit tests. I have the lua scripts worked out to temporarily set the cursor color, but I'm not sure how best to invoke those lua functions at the end of the unit testing task.

I set up a keymap to run `require('overseer').run_template({name = 'unit_test'})`, but it looks like anything I put after that runs before the task completes (nor do I know how I would get context of the task result even if the task had completed at this point). I thought I might be on to something when I saw there is an option to pass a callback function to the run_template, but if the `task` argument's status being `"RUNNING"` is any indication that the task has not completed yet, it looks like this callback isn't called when the task stops running?

Any ideas would be greatly appreciated!

**Comment by stevearc** [comment, marked as answer]
There are a few ways to do this, but the one I would recommend is to create a [custom component](https://github.com/stevearc/overseer.nvim/blob/master/doc/guides.md#custom-components) and add it to your task in the template definition.

**Reply by georgev93** [reply]
Ah. Looks like that's the sort of customization that components allow! Thanks for your help!

I'll just post my complete component here in case someone finds themselves here trying to do something similar (trying to be a good samaritan!)

I was able to get this working with the following component addition:

``` lua
-- ~/.config/nvim/lua/overseer/component/test_cursor/test_cursor.lua
local function set_cursor_color(color)
  vim.cmd('highlight testCursor guibg=' .. color)
end

local function reset_cursor_color()
  vim.cmd('highlight testCursor guibg=NONE')
  vim.opt.guicursor = nil
end

return {
  desc = 'This component changes the cursor color based on the result of a test',
  constructor = function(params)
    return {
      on_start = function(self, task)
        set_cursor_color('yellow')
        vim.opt.guicursor = 'a:testCursor'
      end,
      on_complete = function(self, task, status, result)
        if status == 'SUCCESS' then
          set_cursor_color('green')
        else
          set_cursor_color('red')
        end
        vim.defer_fn(reset_cursor_color, 10000)
      end,
    }
  end,
}
```

Since I don't normally have guicursor set to anything, this will override my cursor color with yellow while my unit tests are running, then change the color of the cursor to red or green depending on the unit test result. After 10 seconds, the cursor goes back to normal. Then I just set up a custom task for my unit test which utilizes this custom component (which I called `test_cursor`):

```lua
-- ~/.config/nvim/lua/overseer/template/user/meson_c_test.lua
return {
  name = 'meson_c_test',
  builder = function()
    local build_path = vim.fn.getcwd() .. '/build'
    return {
      cmd = { 'ninja' },
      args = { '-C', build_path, 'unit_test' },
      components = {
        'default',
        'test_cursor.test_cursor',
      },
    }
  end,
}
```
